monitoring/prometheus/prometheus.go

```diff
@@ -0,0 +1,195 @@
+	prom.MustRegister(serverStreamMsgReceived)
+	prom.MustRegister(serverStreamMsgSent)
+	prom.MustRegister(serverHandledHistogram)
+	prom.MustRegister(serverErred)
+	prom.MustRegister(clientStartedCounter)
+	prom.MustRegister(clientStreamMsgReceived)
+	prom.MustRegister(clientStreamMsgSent)
+	prom.MustRegister(clientHandledHistogram)
+	prom.MustRegister(clientErred)
+}
+
+type ServerMonitor struct {
+}
+
+func (m *ServerMonitor) NewServerMonitor(rpcType monitoring.RpcType, fullMethod string) monitoring.RpcMonitor {
+	r := &serverRpcMonitor{rpcType: rpcType, startTime: time.Now()}
+	r.serviceName, r.methodName = splitMethodName(fullMethod)
+	serverStartedCounter.WithLabelValues(string(r.rpcType), r.serviceName, r.methodName).Inc()
+	return r
+}
+
+type serverRpcMonitor struct {
+	rpcType     monitoring.RpcType
+	serviceName string
+	methodName  string
+	startTime   time.Time
+}
+
+func (r *serverRpcMonitor) ReceivedMessage() {
+	serverStreamMsgReceived.WithLabelValues(string(r.rpcType), r.serviceName, r.methodName).Inc()
+}
+
+func (r *serverRpcMonitor) SentMessage() {
+	serverStreamMsgSent.WithLabelValues(string(r.rpcType), r.serviceName, r.methodName).Inc()
+}
+
+func (r *serverRpcMonitor) Handled(code codes.Code) {
+	serverHandledHistogram.WithLabelValues(string(r.rpcType), r.serviceName, r.methodName, code.String()).Observe(time.Since(r.startTime).Seconds())
+}
+
+func (r *serverRpcMonitor) Erred(err error) {
+	serverErred.WithLabelValues(string(r.rpcType), r.serviceName, r.methodName, errorType(err)).Inc()
+}
+
+type ClientMonitor struct {
+}
+
+func (m *ClientMonitor) NewClientMonitor(rpcType monitoring.RpcType, fullMethod string) monitoring.RpcMonitor {
+	r := &clientRpcMonitor{rpcType: rpcType, startTime: time.Now()}
+	r.serviceName, r.methodName = splitMethodName(fullMethod)
+	clientStartedCounter.WithLabelValues(string(r.rpcType), r.serviceName, r.methodName).Inc()
+	return r
+}
+
+type clientRpcMonitor struct {
+	rpcType     monitoring.RpcType
+	serviceName string
+	methodName  string
+	startTime   time.Time
+}
+
+func (r *clientRpcMonitor) ReceivedMessage() {
+	clientStreamMsgReceived.WithLabelValues(string(r.rpcType), r.serviceName, r.methodName).Inc()
+}
+
+func (r *clientRpcMonitor) SentMessage() {
+	clientStreamMsgSent.WithLabelValues(string(r.rpcType), r.serviceName, r.methodName).Inc()
+}
+
+func (r *clientRpcMonitor) Handled(code codes.Code) {
+	clientHandledHistogram.WithLabelValues(string(r.rpcType), r.serviceName, r.methodName, code.String()).Observe(time.Since(r.startTime).Seconds())
+}
+
+func (r *clientRpcMonitor) Erred(err error) {
+	clientErred.WithLabelValues(string(r.rpcType), r.serviceName, r.methodName, errorType(err)).Inc()
+}
+
+func splitMethodName(fullMethodName string) (string, string) {
+	fullMethodName = strings.TrimPrefix(fullMethodName, "/") // remove leading slash
+	if i := strings.Index(fullMethodName, "/"); i >= 0 {
+		return fullMethodName[:i], fullMethodName[i+1:]
```

Comment:
Monitoring on service + method name allows you to do richer queries in Prometheus. E.g. count all errors in a given Service, regardless of Method.